### PR TITLE
🐛 ClusterResourceSet: continue applying when apply for a single cluster failed

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:
Previously as soon as apply for a single cluster failed the Reconcile func returned and other clusters were not reconciled.

With this PR we try to apply for all clusters and then requeue if one of them returned an `ErrClusterLocked` error.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by sbueringer** [comment]
/hold 
I think we might not need this

**Comment by sbueringer** [comment]
Looks like we don't need this fix once this CR fix is merged: https://github.com/kubernetes-sigs/controller-runtime/pull/2303

Basically with the old CR version ConfigMap/Secret were not cached, with the new version they were cached. So we now were hitting stale cache issues with ConfigMap/Secret.

**Comment by killianmuldoon** [comment]
Should we fix this anyway? I think this PR was all upside / improvements in CRS.

**Comment by sbueringer** [comment]
Do you think we should requeue if the object actually does not exist?

**Comment by sbueringer** [comment]
> Do you think we should requeue if the object actually does not exist?

The part that we try all clusters before we return the errors is definitely an improvement we should merge

File: exp/addons/internal/controllers/clusterresourceset_controller.go
Comment (fabriziopandini) [inline exp/addons/internal/controllers/clusterresourceset_controller.go:298]: Q: Should we add a comment explaining why we are requeing resources not found?

File: exp/addons/internal/controllers/clusterresourceset_controller.go
Comment (fabriziopandini) [inline exp/addons/internal/controllers/clusterresourceset_controller.go:373]: Q: should this go before setting the condition to true?

File: exp/addons/internal/controllers/clusterresourceset_controller.go
Comment (sbueringer) [reply exp/addons/internal/controllers/clusterresourceset_controller.go:298]: In general yes.

The question is if we still want to do it.

Initially I added this because through a bug in the new CR version I was hitting a stale cache with get CM/Secret. This doesn't happen anymore since I fixed CR.

Now the only reason to requeue is basically to get a quick requeue if someone deploys first the CRS and then the referenced CM/Secret.

WDYT should we requeue for those cases?

If we don't the next resync will catch it.

File: exp/addons/internal/controllers/clusterresourceset_controller.go
Comment (sbueringer) [reply exp/addons/internal/controllers/clusterresourceset_controller.go:373]: As discussed move this up a bit

**Comment by sbueringer** [comment]
I'll update the PR as discussed.

Independent of that I want to merge this 1-2 weeks after the CR bump to ensure the CR bump by itself is not introducing a regression

/hold

File: exp/addons/internal/controllers/clusterresourceset_controller.go
Comment (sbueringer) [reply exp/addons/internal/controllers/clusterresourceset_controller.go:298]: Thougt a bit more about it. The requeue doesn't make sense (now that we are not caching CM/Secret anymore).

The controller is watching on all ConfigMaps / Secrets, so we don't need a requeue for the case a ConfigMap / Secret is created later on.

**Comment by sbueringer** [comment]
@fabriziopandini Reframed the PR (explanation here: https://github.com/kubernetes-sigs/cluster-api/pull/8611#discussion_r1187677215).

I think now we can also merge this PR independent of the CR bump. I don't see a relevant overlap anymore

/assign @fabriziopandini 
/assign @g-gaston

**Review by g-gaston — COMMENTED** [review]
thanks for this @sbueringer 
/lgtm

**Review by vincepri — APPROVED** [review]
/approve

**Comment by fabriziopandini** [comment]
/lgtm
/approve

ok to hold as you proposed

**Comment by sbueringer** [comment]
@fabriziopandini I would merge now directly because of https://github.com/kubernetes-sigs/cluster-api/pull/8611#issuecomment-1538744223 .

Basically this PR doesn't change the reqeue behavior for a CRS with missing resources anymore.

/hold cancel

**Pull request merged**
